src/blob.rs

```diff
@@ -0,0 +1,162 @@
+                            actual: digest,
+                        },
+                    ))))
+                }
+            }
+        }
+    }
+}
+
+/// Helper wrapper around various digest algorithms to make it easier to use them with our blob
+/// utils. This has to be an enum because the digest algorithms aren't object safe so we can't box
+/// dynner them
+pub(crate) enum Digester {
+    Sha256(sha2::Sha256),
+    Sha384(sha2::Sha384),
+    Sha512(sha2::Sha512),
+}
+
+impl Digester {
+    pub fn new(digest: &str) -> crate::errors::Result<Self> {
+        let (algo, _) = digest.split_once(':').ok_or_else(|| {
+            OciDistributionError::GenericError(Some(format!(
+                "Digest header value is not in the expected format: {}",
+                digest
+            )))
+        })?;
+
+        match algo {
+            "sha256" => Ok(Digester::Sha256(sha2::Sha256::new())),
+            "sha384" => Ok(Digester::Sha384(sha2::Sha384::new())),
+            "sha512" => Ok(Digester::Sha512(sha2::Sha512::new())),
+            _ => Err(OciDistributionError::GenericError(Some(format!(
+                "Unsupported digest algorithm: {}",
+                algo
+            )))),
+        }
+    }
+
+    pub fn update(&mut self, data: impl AsRef<[u8]>) {
+        match self {
+            Self::Sha256(d) => d.update(data),
+            Self::Sha384(d) => d.update(data),
+            Self::Sha512(d) => d.update(data),
+        }
+    }
+
+    pub fn finalize(&mut self) -> String {
+        match self {
+            Self::Sha256(d) => format!("sha256:{:x}", d.finalize_reset()),
+            Self::Sha384(d) => format!("sha384:{:x}", d.finalize_reset()),
+            Self::Sha512(d) => format!("sha512:{:x}", d.finalize_reset()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use bytes::Bytes;
+    use futures_util::TryStreamExt;
+
+    #[tokio::test]
+    async fn test_verifying_stream() {
+        // Test with correct SHA
+        let data = b"Hello, world!";
+        let correct_sha = format!("sha256:{:x}", sha2::Sha256::digest(data));
+        let stream = VerifyingStream::new(
+            Box::pin(futures_util::stream::iter(vec![Ok(Bytes::from_static(
+                data,
+            ))])),
+            Digester::new(&correct_sha).unwrap(),
+            correct_sha,
+        );
+        stream
+            .try_collect::<Vec<_>>()
+            .await
+            .expect("Should not error with valid data");
+
+        // Test with incorrect SHA
+        let incorrect_sha = "sha256:incorrect_hash";
+        let stream = VerifyingStream::new(
+            Box::pin(futures_util::stream::iter(vec![Ok(Bytes::from_static(
+                data,
+            ))])),
+            Digester::new(incorrect_sha).unwrap(),
+            incorrect_sha.to_string(),
+        );
+        stream
+            .try_collect::<Vec<_>>()
+            .await
+            .expect_err("Should error with invalid sha");
```

Comment:
By using a dedicated error type to identify the invalid digest format, we could write a more specific assert. We could make sure the returned error is of the type we expected (`using `matches!`)